### ISSUE TITLE
Octuple decode KeyError

### ISSUE THREAD

**Issue by NHeLv1** [Natooz/MidiTok#206]
I trained an autoregressive Transformer using Octuple, but a KerError occurs when it eventually decodes into midi

**Comment by NHeLv1** [comment]
![image](https://github.com/user-attachments/assets/bd723de7-2c1d-4a18-b4f3-f30aac1bc5b9)

**Comment by Natooz** [comment]
Hi,
Can you share more information on the tokenizer (ideally its configuration, as a tokenizer.json file for example) and the token ids for which this error is occurring?

**Comment by NHeLv1** [comment]
I think I've understood something.When using the octuple tokenizer,I didn't retrieve all the ids for each of its tracks,and the size of the embeddings my model uses exceeded the id range,which is why a KeyError occurred.So,where can I obtain the octuple's ids?

**Comment by NHeLv1** [comment]
I change my Transformer into multi vocab, but I got a new problem
![image](https://github.com/user-attachments/assets/b9b6d2fa-63ca-4965-86d4-ab6d2e0305f6)
and Here is my code 
`
class NaiveGPT(nn.Module):
    def __init__(self, config: GPT1Config, multi_vocab_lens=None):
        super().__init__()
        self.config = config
        # self.wte = nn.Embedding(config.vocab_size, config.d_model) # word token embeddings
        self.wpe = PositionalEncoding(config.max_len, config.d_model) # word position encodings
        self.blocks = nn.ModuleList([DecoderBlock(config) for _ in  range(config.n_layers)])
        # self.linear1 = nn.Linear(config.d_model, config.vocab_size)

        self.wte = nn.ModuleList([nn.Embedding(i, config.d_model) for i in multi_vocab_lens])
        self.linear2vocab =  nn.ModuleList([nn.Linear(config.d_model, i) for i in multi_vocab_lens])

        # self.wte.weight = self.linear1.weight
        
        self.projection_in = nn.Linear(config.d_model * len(multi_vocab_lens), config.d_model)
        self.projection_out = nn.Linear(config.d_model, config.d_model * len(multi_vocab_lens))

    def forward(self, inputs, mask, targets = None):
        bs, sl, track_num = inputs.shape
        inputs = inputs.permute(2, 0, 1)
        embedded = [embed(token) for embed, token in zip(self.wte, inputs)]
        concat_embed = torch.cat(embedded, dim=-1)

        # logits = self.wte(inputs) # dim -> batch_size, sequence_length, d_model
        logits = self.projection_in(concat_embed)
        logits = self.wpe(logits)
        for block in self.blocks:
            logits = block(logits, mask)
        # logits = self.linear1(logits)
        logits = self.projection_out(logits)

        logits = logits.view(bs, sl, -1, track_num).permute(3, 0, 1, 2)
        

        out_tokens = [linear(logit) for linear, logit in zip(self.linear2vocab, logits)]
    
        # out_tokens = torch.cat(out_tokens, dim = -1) # 每个vocab大小不一样时不能cat

        loss_ce = None
        if targets != None:
            targets = targets.permute(2, 0, 1)
            
            for i, (tgt_track, out_track) in enumerate(zip(targets, out_tokens)):
                
                batch_size, sequence_length, track_vocab_size = out_track.shape
                out_track = out_track.view(batch_size * sequence_length, track_vocab_size)
                tgt_track = tgt_track.view(batch_size * sequence_length)
                # print(out_track.shape)
                # print(tgt_track.shape)

                loss_i = F.cross_entropy(out_track, tgt_track, ignore_index=-100)
                if i == 0:
                    loss_ce = loss_i
                else:
                    loss_ce += loss_i


            """ 
            batch_size, sequence_length, vocab_size, track_num = out_tokens.shape
            # to calculate loss for all token embeddings in a batch
            # kind of a requirement for cross_entropy
            out_tokens = out_tokens.permute(3, 0, 1, 2).reshape(batch_size * sequence_length * track_num, vocab_size)
            targets = targets.view(batch_size * sequence_length * track_num)
            # print(targets.shape)
            valid_positions = targets != -100
            out_tokens = out_tokens[valid_positions]
            targets = targets[valid_positions]
            """
            # loss = F.cross_entropy(out_tokens, targets, ignore_index=-100)
        # raise ValueError("coding")
        return out_tokens, loss_ce
    

    def generate(self, inputs, mask, max_new_tokens, tokenizer):
        
        output = inputs.clone()
        for _ in range(max_new_tokens):
            current_seq_length = inputs.size(1)
            # Truncate inputs if it exceeds context_length
            if current_seq_length > self.config.max_len:
                inputs = inputs[:, -self.config.max_len:]
            # we only pass targets on training to calculate loss
            
            mask = torch.ones_like(inputs.float().mean(dim = -1))
            # print(inputs.shape, mask.shape)
            logits, _ = self(inputs, mask)  
            
            # multi track modified
            # print(inputs.shape)
            idx_next = []
            for i, track in enumerate(logits):
                logits = track[:, -1, :] 
                # print(logits.shape)
                probs = F.softmax(logits, dim=1)
                track_idx_next = torch.multinomial(probs, num_samples=1)
                idx_next.append(track_idx_next)
                # print(track_idx_next.shape)
            idx_next = torch.cat(idx_next, dim=-1)
            idx_next = idx_next.unsqueeze(1)
            # print(idx_next.shape)

            inputs = torch.cat([inputs, idx_next], dim=1)
            output = torch.cat([output, idx_next], dim=1)
            # print(inputs.shape)

            # raise ValueError('coding')
        
        for out in output:
            print(out.tolist())
        return [tokenizer(out.tolist()) for out in output]
`

**Comment by Natooz** [comment]
(Apologies for the late reply)

Indeed Octuple is "multi-vocabulary", i.e. the model needs multiple embedding layers, and multiple output modules (one per token type).
The last error seems to occur because the model predicted a non-valid time signature token (probably "something"_None within this sub vocabulary). I don't how how is the model trained exactly so I can't tell if it is expected for the model to predict any token with a "None" value part (EOS/BOS?). In the case of EOS, I suggest to preprocess the token ids and cut the sequence when one is met.

**Comment by Natooz** [comment]
I indeed do not recommend Octuple for generation tasks, for several reasons. First having multiple vocabularies/input modules/output modules increases the complexity of the model, and of the training procedure that now musts compute several losses, this must be implemented and tested and might not be compatible with popular libraries such assign HF transformers. Secondly sampling multiple tokens from different distributions that are not computed with any of them conditioning the others leads to higher variance and potentially incoherent results. That's less of an issue for bidirectional transformers (i.e. no attention masking) as the prediction error doesn't accumulate over autoregressive generation steps. Third, BPE/Unigram is to me a much better alternative to reduce the sequence length, almost matching octuple's reduction ratio (can surpass it depending on the data/vocab size), is compatible out of the box with all models/libraries and allows to benefit from increases "semantic"/information density represented by the aggregated tokens in the vocabulary.
